main.go

```diff
@@ -197,6 +200,141 @@ type MainController struct {
 	db *sqlx.DB
 }
 
+func (c MainController) authAttendanceMiddleware(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		user, authed := getAuthedADBUser(c.db, r)
+		if !authed {
+			// Delete the cookie if it doesn't auth.
+			c := &http.Cookie{
+				Name:     "auth-session",
+				Path:     "/",
+				MaxAge:   -1,
+				HttpOnly: true,
+			}
+			http.SetCookie(w, c)
+
+			http.Redirect(w, r, "/login", http.StatusFound)
+			return
+		}
+
+		allowedRoles := []string{"attendance", "organizer", "admin"}
+		fmt.Println(user)
+		if !userIsAllowed(allowedRoles, user) {
+			fmt.Println("NOT ALLOWED!")
+			http.Redirect(w, r.WithContext(setUserContext(r, user)), "/403", http.StatusFound)
+			return
+		}
+		fmt.Println("Allowing Access")
+		// Request is authed at this point.
+		h.ServeHTTP(w, r.WithContext(setUserContext(r, user)))
+	})
+}
+
+func (c MainController) authOrganizerMiddleware(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		user, authed := getAuthedADBUser(c.db, r)
+		if !authed {
+			// Delete the cookie if it doesn't auth.
+			c := &http.Cookie{
+				Name:     "auth-session",
+				Path:     "/",
+				MaxAge:   -1,
+				HttpOnly: true,
+			}
+			http.SetCookie(w, c)
+
+			http.Redirect(w, r, "/login", http.StatusFound)
+			return
+		}
+
+		allowedRoles := []string{"organizer", "admin"}
+		fmt.Println(user)
+		if !userIsAllowed(allowedRoles, user) {
+			fmt.Println("NOT ALLOWED!")
+			http.Redirect(w, r.WithContext(setUserContext(r, user)), "/403", http.StatusFound)
+			return
+		}
+		fmt.Println("Allowing Access")
+		// Request is authed at this point.
+		h.ServeHTTP(w, r.WithContext(setUserContext(r, user)))
+	})
+}
+
+func (c MainController) authAdminMiddleware(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		user, authed := getAuthedADBUser(c.db, r)
+		if !authed {
+			// Delete the cookie if it doesn't auth.
+			c := &http.Cookie{
+				Name:     "auth-session",
+				Path:     "/",
+				MaxAge:   -1,
+				HttpOnly: true,
+			}
+			http.SetCookie(w, c)
+
+			http.Redirect(w, r, "/login", http.StatusFound)
+			return
+		}
+
+		allowedRoles := []string{"admin"}
+
+		if !userIsAllowed(allowedRoles, user) {
+			http.Redirect(w, r.WithContext(setUserContext(r, user)), "/403", http.StatusFound)
+			return
+		}
+
+		// Request is authed at this point.
+		h.ServeHTTP(w, r.WithContext(setUserContext(r, user)))
+	})
+}
+
+func userIsAllowed(roles []string, user model.ADBUser) bool {
+	if len(roles) == 0 {
+		return false
+	}
+
+	if len(user.Roles) == 0 {
+		return false
+	}
+
+	var allowed = false
+
+	for i := 0; i < len(roles); i++ {
```

Comment:
I'd get rid of the two `if` statements above, and get rid of `allowed` by using just `return true` in the loop, and `return false` at the end.

Reply:
Sure, I can do that. I generally don't mind having sanity checks in front of any operations, and early returns. Seems like half of the time I'd do it this way, and the other half the way you described.

Reply:
A lot of software engineering studies report number of issues is linear with lines of code, so I generally favor omitting special case logic if they can be handled by generic code without extra work or a performance hit.

Reply:
It also helps that Go is a strongly typed language. I think in Javascript sanity checks are more crucial because of non-typed variables.

Also, don't know why this isn't showing up as "Outdated" because I took your advice. Smh.